identityRatio <- function(aln) {
    seqs <- sapply(aln$seqs[,2], strsplit, split="")
    sum(seqs[[1]] == seqs[[2]])/ aln$length
}

getIdx <- function(queryID, sourceID) {
    unlist(sapply(queryID, function(x) which(x == sourceID)))
}

pause <- function() {
    print("press ENTER to continue...")
    readline()
    invisible()
}

aln2seqDF <- function(aln) {
    seqs <- aln$seqs[,2]
    seq.df <- do.call("rbind", strsplit(seqs, ""))
    return(seq.df)
}

printFormatSeq <- function(seq, window=80) {
    jj <- seq(1, nchar(seq), window)
    if (jj[length(jj)] < nchar(seq)) {
        jj <- c(jj, nchar(seq))
    }
    for (p in 1:(length(jj)-1)) {
        cat(substring(seq, jj[p], jj[p+1]), "\n")
    }
}


addFooter <- function(file) {
    sink(file, append=TRUE)
    cat("\n")
    cat("## Info\n")
    cat("\nThis Report was generated by `skleid` version", as.character(packageVersion("skleid")),
        ".\nContact [Guangchuang](mailto:gcyu@connect.hku.hk) if you need helps.\n")
    sink() 
}

##' print Info
##'
##' 
##' @title printInfo
##' @return NULL
##' @author ygc
##' @export
printInfo <- function() {
    cat("\n###################################################################\n")
    cat("##\t\t\t\t\t\t\t\t ##\n")
    cat("##\t\t\t\t\t\t\t\t ##\n")
    cat('## Author: Guangchuang Yu (gcyu@connect.hku.hk)\t\t\t ##\n')
cat(paste('## skleid package, version ',
          as.character(packageVersion("skleid")), ', was installed', sep=""),
    "\t\t ##\n")
    cat('## This package designed for SKLEID lab for internal used.\t ##\n')
    cat('## use command, help(package="skleid"), to view online manuals\t ##\n') 
    cat('## If you got any problem, please contact me by email\t\t ##\n')
    cat("##\t\t\t\t\t\t\t\t ##\n")
    cat("##\t\t\t\t\t\t\t\t ##\n")
    cat("###################################################################\n\n")

}

getFiles <- function(path) {
    ff <- list.files(path=path)
    ff <- paste(path, ff, sep="/")
    return(ff)
}

doFilter <- function(contig.folder, percentage) {
    contig <- getFiles(contig.folder)
     
    f454 <- contig[grep("_454.fa[sta]$", contig)] ## contig[grep("454", contig)]
    cat(">>", "filtering sequence read numbers below", percentage, "percentage", "\n") 
    readNumFilter(f454, percentage)
}

moveUnknownFile <- function(contig.folder) {
    contig <- getFiles(contig.folder)
    sc <- getSampleID(contig)
    idx <- which(sc == contig)
    if (length(idx) >= 1) {
        moveFile(contig[idx], contig.folder, "unknown")
    }
}

unKnownFileReport <- function(contig.folder, report.file) {
    ## move unknown files if any
    moveUnknownFile(contig.folder)
    if (file.exists("unknown")) {
        num <- length(list.files(path="unknown"))
        if (num > 0) {
            sink(report.file, append=TRUE)
            cat(num, " [unknown](unknown) file(s) found.\n")
            sink()
            cat(">>", num, " unknown file(s) found.\n")
        }
    }
}

moveFile <- function(files, from, to) {
    if (!file.exists(to))
        dir.create(to)
    for (ff in files) {
        file.rename(ff, to=sub(from, to, ff))
    }
}

moveEmptyFile <- function(folder) {
    ff <- getFiles(folder)
    ii <- which(file.info(ff)$size == 0)
    if (length(ii) > 0) {
        moveFile(ff[ii], folder, "empty")
    }
}

emptyFileReport <- function(contig.folder, ref.folder, report.file) {
    moveEmptyFile(contig.folder)
    moveEmptyFile(ref.folder)

    if (file.exists("empty")) {
        num <- length(list.files(path="empty"))
        if (num > 0) {
            sink(report.file, append=TRUE)
            cat(num, " [empty](empty) file(s) found.\n")
            sink()
            cat(">>", num, " empty file(s) found.\n")
        }
    }
}

getMixedFileIndex <- function(files) {
    which(sapply(files, isMixed))
}


isMixed <- function(file) {
    ## prot <- gsub(pattern=".*/[^SRL]*[SRL]+\\d+(\\w+\\d*)_454.*", replacement="\\1", file)
    prot <- gsub(".*/.*_[SRL]+\\d+([HNMP][APSB]\\d*)_.*", replacement="\\1", file)
    
    prot.cutoff <- c(rep(2000, 4), 1000, rep(3000, 3))
    names(prot.cutoff) <- c("HA", "NA", "MP", "NP", "NS", "PA", "PB1", "PB2")
    
    res <- file.info(file)$size > prot.cutoff[prot]
    ## NDV that is not exists
    if (is.na(res))
        return(TRUE)
    return(res)
}

getSampleID <- function(contig) {
    sc <- gsub(".*/.*_([SRL]+\\d+)[HNMP]+.*", replacement="\\1", contig)
    return(sc)
}


moveMixedFile <- function(contig.folder) {
    contig <- getFiles(contig.folder)
    f454 <- contig[grep("_454.fa[sta]$", contig)]
    sc <- getSampleID(contig)
    idx <- getMixedFileIndex(f454)
    
    mix <- unique(gsub(pattern=".*/.*_([SRL]+\\d+)[HNMP][APSB]\\d*_.*", replacement="\\1", f454[idx]))
        
    if (length(mix) >= 1) {
        if (!file.exists("mixed"))
            dir.create("mixed")

        ii <- which(sc %in% mix)
        moveFile(contig[ii], contig.folder, "mixed")
    }

}

getMixedStrain <- function() {
    mix.sc <- NULL
    if (file.exists("mixed")) {
        mix <- getFiles("mixed")
        if (length(mix) > 0)
            mix.sc <- getSampleID(mix)
        if (length(mix.sc) == 0)
            mix.sc <- NULL
    }
    return(mix.sc)
}

mixedFileReport <- function(contig.folder, report.file) {
    mix.sc <- getMixedStrain()  
    if ( !is.null(mix.sc)) {
        sink(report.file, append=TRUE)
        cat("\n", length(unique(mix.sc)), " [mixed](mixed) strain(s) found.\n")
        cat("\n\t", paste(unique(mix.sc)), "\n")
        sink()
        cat(">>", length(unique(mix.sc)), " mixed strain(s) found.\n")
    }
    
    contig <- getFiles(contig.folder)
    sc <- getSampleID(contig)
    if (length(sc) > 0) {
        cat(">>", length(unique(sc)), " strains were processed in the report.\n")
        ## processing 
        sink(report.file, append=TRUE)
        cat(length(unique(sc)), " strains were processed in the report.\n")
        cat("\n\t", paste(unique(sc)), "\n\n")
        sink()
    }
}


generateStrainTable <- function(contig.folder, nameMap) {
    mix.sc <- getMixedStrain()
    if (!is.null(mix.sc)) {
        mixff <- nameMap[ nameMap[,1] %in% sub("[SRL]+", "", unique(mix.sc)),]    
        mixff[,2] <- paste(mixff[,2], "mixed", sep="_")
        write.table(mixff,
                    file="mixed_table.txt", sep="\t",
                    row.names=F, col.names=F, quote=F)        
    }
        
    contig <- getFiles(contig.folder)
    sc <- getSampleID(contig)
    write.table(nameMap[nameMap[,1] %in% sub("S", "", unique(sc)), ],
                file="unmixed_table.txt", sep="\t",
                row.names=F, col.names=F, quote=F)        
}

moveNDVFile <- function(contig.folder) {
    ## S38-Yellow-CK-JX-10226-2014_S38NDV_454.fas
    ## S38-Yellow-CK-JX-10226-2014_S38NDV_Contigs.fna
    ## S38-Yellow-CK-JX-10226-2014_S38NDV_mira_delX.fasta
    ## S38-Yellow-CK-JX-10226-2014_S38NDV_mira.fasta
    contig <- getFiles(contig.folder)
    
    idx <- grep(".*/*_[SRL]+\\d+NDV_.*", contig)
    
    if (length(idx) >= 1) {
        if (!file.exists("NDV"))
            dir.create("NDV")
        moveFile(contig[idx], contig.folder, "NDV")
    }
}

NDVFileReport <- function(contig.folder, report.file) {
    moveNDVFile(contig.folder)

    if (file.exists("NDV")) {
        ff <- list.files(path="NDV")
        if (length(ff) > 0) {
            ndv <- gsub(".*/*_([SRL]+\\d+NDV)_.*",'\\1', ff)
            ndv <- unique(ndv)
            sink(report.file)
            cat("\n", length(ndv), " [NDV](NDV) gene(s) found.\n")
            cat("\n\t", paste(ndv), "\n")
            sink()
        }
    }
}
